### PR TITLE
drivers: can: support for Nuvoton numaker can

### PR DESCRIPTION
**Pull request by cyliangtw**
This PR is to add Nuvoton numaker series can controller.
It's verified on board: NuMaker-PFM-M467.

### CODE REVIEW

File: drivers/can/can_numaker.c [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+	uint32_t    clk_div;
+	const struct device *clk_dev;
+	void (*irq_config_func)(const struct device *dev);
+	const struct pinctrl_dev_config *pincfg;
+};
+
+struct can_numaker_data {
+	/* Avoid zero length struct */
+	uint32_t resv0;
+};
```
Comment (henrikbrixandersen) [inline drivers/can/can_numaker.c:43]: You can just leave out the struct. No need for it.

File: drivers/can/can_numaker.c [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+	}
+	clksrc_rate = clksrc_rate_tab[clksrc_rate_idx];
+
+	/* Module clock divider */
+	clkdiv_divider = CLK_GetModuleClockDivider(config->clk_modidx) + 1;
+
+	LOG_DBG("Clock source/divider: %d/%d", clksrc_rate, clkdiv_divider);
+	*rate = clksrc_rate / clkdiv_divider;
+
+cleanup:
```
Comment (henrikbrixandersen) [inline drivers/can/can_numaker.c:90]: Label is misleading. No cleaning is performed here.

File: drivers/can/can_numaker.c [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+	SYS_UnlockReg();
+
+	memset(&scc_subsys, 0x00, sizeof(scc_subsys));
+	scc_subsys.subsys_id        = NUMAKER_SCC_SUBSYS_ID_PCC;
+	scc_subsys.pcc.clk_modidx   = config->clk_modidx;
+	scc_subsys.pcc.clk_src      = config->clk_src;
+	scc_subsys.pcc.clk_div      = config->clk_div;
+
+	/* Equivalent to CLK_EnableModuleClock() */
```
Comment (henrikbrixandersen) [inline drivers/can/can_numaker.c:117]: Why are these comments here? These are Zephyr APIs. No need to mention what they would be equivalent to in some other API.

File: drivers/can/can_numaker.c [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+	if (rc < 0) {
+		LOG_ERR("Failed to initialize mcan: %d", rc);
+		goto cleanup;
+	}
+
+	rc = can_numaker_get_core_clock(dev, &rate);
+	if (rc < 0) {
+		goto cleanup;
+	}
+	LOG_INF("CAN core clock: %d", rate);
```
Comment (henrikbrixandersen) [inline drivers/can/can_numaker.c:154]: Please avoid INF logging in driver init functions. Also, why is this needed? Looks like left-in debugging.

File: drivers/can/can_numaker.c [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+	.get_max_bitrate = can_mcan_get_max_bitrate,
+	/* Nominal bit timing limits
+	 *
+	 * NUMAKER MCAN timing limits are specified in the "Nominal bit timing and
+	 * prescaler register (NBTP)" table in the SoC reference manual.
+	 *
+	 * NOTE: The values here are the "physical" timing limits, whereas the register
+	 *       field limits are physical values minus 1 (which is handled by the
+	 *       register assignments in the common MCAN driver code).
+	 */
```
Comment (henrikbrixandersen) [inline drivers/can/can_numaker.c:187]: This is standard for all Bosch M_CAN IPs. No need for this comment.

File: drivers/can/can_numaker.c [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+	.timing_max = CAN_MCAN_TIMING_MAX_INITIALIZER,
+#ifdef CONFIG_CAN_FD_MODE
+	.set_timing_data = can_mcan_set_timing_data,
+	/* Data bit timing limits
+	 *
+	 * NUMAKER MCAN timing limits are specified in the "Data bit timing and
+	 * prescaler register (DBTP)" table in the SoC reference manual.
+	 *
+	 * NOTE: Same "minus one" concern as above.
+	 */
```
Comment (henrikbrixandersen) [inline drivers/can/can_numaker.c:198]: Same here.

File: drivers/can/can_numaker.c [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+	.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),
+
+#define CAN_NUMAKER_INIT(inst)                  \
+	NUMAKER_PINCTRL_DEFINE(inst);               \
+	CAN_MCAN_DT_INST_CALLBACKS_DEFINE(inst, can_numaker_cbs_##inst);	    \
+												\
+	static void can_numaker_irq_config_func_##inst(const struct device *dev);         \
+												\
+	static const struct can_numaker_config can_numaker_config_##inst = {              \
+		.canfd_base = DT_INST_REG_ADDR_BY_NAME(inst, m_can),                \
```
Comment (henrikbrixandersen) [inline drivers/can/can_numaker.c:268]:
```suggestion
		.canfd_base =CAN_MCAN_DT_INST_MCAN_ADDR(inst),                \
```

File: drivers/can/can_numaker.c [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+		.mrba = CAN_MCAN_DT_INST_MRBA(inst),			    \
+		.mram = CAN_MCAN_DT_INST_MRAM_ADDR(inst),			    \
+		.reset = RESET_DT_SPEC_INST_GET(inst),                                             \
+		.clk_modidx = DT_INST_CLOCKS_CELL(inst, clock_module_index),    \
+		.clk_src    = DT_INST_CLOCKS_CELL(inst, clock_source),          \
+		.clk_div    = DT_INST_CLOCKS_CELL(inst, clock_divider),         \
```
Comment (henrikbrixandersen) [inline drivers/can/can_numaker.c:274]: Too many spaces here.

File: drivers/can/can_numaker.c [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+			&can_numaker_config_##inst,      \
+			&can_numaker_ops,		    \
+			&can_numaker_cbs_##inst);	    \
+											\
+	static struct can_numaker_data can_numaker_data_##inst;           \
+									\
+	static struct can_mcan_data can_mcan_data_##inst =                \
+		CAN_MCAN_DATA_INITIALIZER(&can_numaker_data_ ## inst);                     \
+									\
+	DEVICE_DT_INST_DEFINE(inst,     \
```
Comment (henrikbrixandersen) [inline drivers/can/can_numaker.c:291]:
```suggestion
	CAN_DEVICE_DT_INST_DEFINE(inst,     \
```

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Thanks! Looks good, but a few comments need addressing:

File: drivers/can/can_numaker.c [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+		.reset = RESET_DT_SPEC_INST_GET(inst),                                             \
+		.clk_modidx = DT_INST_CLOCKS_CELL(inst, clock_module_index),    \
+		.clk_src    = DT_INST_CLOCKS_CELL(inst, clock_source),          \
+		.clk_div    = DT_INST_CLOCKS_CELL(inst, clock_divider),         \
+		NUMAKER_CLKCTRL_DEV_INIT(inst)                                  \
+		.irq_config_func    = can_numaker_irq_config_func_##inst,     \
+		NUMAKER_PINCTRL_INIT(inst)                                      \
+	};                                          \
+												\
+	static const struct can_mcan_config can_mcan_config_##inst =      \
```
Comment (fabiobaltieri) [inline drivers/can/can_numaker.c:280]: the `\` here are all over the place with pretty much any tabstop value (and the right one is 8 anyway :-), could you align them?

File: drivers/can/can_numaker.c [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+static const struct can_mcan_ops can_numaker_ops = {
+	.read_reg = can_numaker_read_reg,
+	.write_reg = can_numaker_write_reg,
+	.read_mram = can_numaker_read_mram,
+	.write_mram = can_numaker_write_mram,
+	.clear_mram = can_numaker_clear_mram,
+};
+
+#define NUMAKER_CLKCTRL_DEV_INIT(inst)          \
+	.clk_dev    = DEVICE_DT_GET(DT_PARENT(DT_INST_CLOCKS_CTLR(inst))),
```
Comment (fabiobaltieri) [inline drivers/can/can_numaker.c:254]:
```suggestion
	.clk_dev = DEVICE_DT_GET(DT_PARENT(DT_INST_CLOCKS_CTLR(inst))),
```

File: drivers/can/can_numaker.c [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+
+	return rc;
+}
+
+static int can_numaker_init(const struct device *dev)
+{
+	const struct can_mcan_config *mcan_config = dev->config;
+	const struct can_numaker_config *config = mcan_config->custom;
+	struct numaker_scc_subsys scc_subsys;
+	int rc = 0;
```
Comment (fabiobaltieri) [inline drivers/can/can_numaker.c:100]: no need to initialize this

File: drivers/can/can_numaker.c [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+struct can_numaker_data {
+	/* Avoid zero length struct */
+	uint32_t resv0;
+};
+
+static int can_numaker_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct can_mcan_config *mcan_config = dev->config;
+	const struct can_numaker_config *config = mcan_config->custom;
+	int rc = 0;
```
Comment (fabiobaltieri) [inline drivers/can/can_numaker.c:49]: no need to initialize this

File: drivers/can/Kconfig.numaker [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+config CAN_NUMAKER
+	bool "Nuvoton NuMaker CAN-FD driver"
+	default y
+	select HAS_NUMAKER_CANFD
+	select CAN_MCAN
+	depends on DT_HAS_NUVOTON_NUMAKER_CANFD_ENABLED
+	help
+	  Enables Nuvoton NuMaker CAN-FD driver, using Bosch M_CAN as backend
```
[the remark this answers is not in the document]
Comment (cyliangtw) [reply drivers/can/Kconfig.numaker:13]: The purpose is to reveal, this driver is to utilize can_mcan instead of fully proprietary implementation.
If there is any concern, it's fine to remove this wording.

File: drivers/can/Kconfig.numaker [hunk 15]
```diff
@@ -0,0 +1,13 @@
+# NuMaker CAN(-FD) driver configuration options
+
+# Copyright (c) 2023 Nuvoton Technology Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+config CAN_NUMAKER
+	bool "Nuvoton NuMaker CAN-FD driver"
+	default y
+	select HAS_NUMAKER_CANFD
```
[the remark this answers is not in the document]
Comment (cyliangtw) [reply drivers/can/Kconfig.numaker:9]: can_mcan API is enough in current implementation and there is no patform dependent part of hal BSP's can function, so it's not necessary.
Thanks of your remind.

File: drivers/can/can_numaker.c [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+#include <zephyr/drivers/clock_control/clock_control_numaker.h>
+#include <zephyr/logging/log.h>
+#include <soc.h>
+#include <NuMicro.h>
+
+LOG_MODULE_REGISTER(can_numaker, CONFIG_CAN_LOG_LEVEL);
+
+/* Implementation notes
+ * 1. Use Bosch M_CAN driver (m_can) as backend
+ *    NOTE: m_can only supports single configuration of max element numbers for all instances
```
[the remark this answers is not in the document]
Comment (cyliangtw) [reply drivers/can/can_numaker.c:23]: After check, m_can already fixed it, your point is right.
Thanks.

File: drivers/can/can_numaker.c (hunk 1 above)
Comment (cyliangtw) [reply drivers/can/can_numaker.c:43]: Done

File: drivers/can/can_numaker.c [hunk 17]
```diff
@@ -0,0 +1,316 @@
...
+	 * NOTE: * NOTE: Fix the conditional for new SOC support
+	 */
+	if (!IS_ENABLED(CONFIG_SOC_SERIES_M46X)) {
+		LOG_ERR("Not support SOC: " CONFIG_SOC_SERIES);
+		rc = -ENOTSUP;
+		goto cleanup;
+	}
+
+	/* Module clock source rate */
+	clksrc_rate_idx = CLK_GetModuleClockSource(config->clk_modidx);
```
Comment (alexanderwachter) [inline drivers/can/can_numaker.c:76]: why are you not using a switch statement here and just call the correct getter instead of filling a table and than pick one?

File: drivers/can/can_numaker.c (hunk 2 above)
Comment (cyliangtw) [reply drivers/can/can_numaker.c:90]: Yes, the Label wording is not correct.
"Done" will be better.

File: drivers/can/can_numaker.c (hunk 3 above)
Comment (cyliangtw) [reply drivers/can/can_numaker.c:117]: The purpose is to record the migration from proprietary CLK control to standard one.
I see and not need mention the proprietary CLK API.

File: drivers/can/can_numaker.c (hunk 4 above)
Comment (cyliangtw) [reply drivers/can/can_numaker.c:154]: Yes, left-in debugging is the better choice.

File: drivers/can/can_numaker.c (hunk 5 above)
Comment (cyliangtw) [reply drivers/can/can_numaker.c:187]: Yes and removed.

File: drivers/can/can_numaker.c [hunk 18]
```diff
@@ -0,0 +1,316 @@
...
+	clksrc_rate_tab[2] = CLK_GetHCLKFreq();
+
+	/* SOC in support list
+	 * NOTE: * NOTE: Fix the conditional for new SOC support
+	 */
+	if (!IS_ENABLED(CONFIG_SOC_SERIES_M46X)) {
+		LOG_ERR("Not support SOC: " CONFIG_SOC_SERIES);
+		rc = -ENOTSUP;
+		goto cleanup;
+	}
```
[the remark this answers is not in the document]
Comment (cyliangtw) [reply drivers/can/can_numaker.c:73]: Yes, handle it in Kconfig.numaker.

File: drivers/can/can_numaker.c (hunk 10 above)
Comment (cyliangtw) [reply drivers/can/can_numaker.c:280]: of course, I aligned them.

File: drivers/can/can_numaker.c (hunk 11 above)
Comment (cyliangtw) [reply drivers/can/can_numaker.c:254]: Fixed it

File: drivers/can/can_numaker.c (hunk 13 above)
Comment (cyliangtw) [reply drivers/can/can_numaker.c:49]: Last time, you guided me the better way.
Applied it: no need variable "rc", return directly as:
```
	if (clksrc_rate_idx >= ARRAY_SIZE(clksrc_rate_tab)) {
		LOG_ERR("Invalid clock source rate table or index");
		return -EIO;
	}
    ......
   return 0;

```

**Comment by cyliangtw** [comment]
Hi @henrikbrixandersen, @fabiobaltieri, @alexanderwachter,
Thanks all of your advice & statement, I updated the code at commit  a103cb0ea7

File: drivers/can/can_numaker.c [hunk 19]
```diff
@@ -0,0 +1,279 @@
...
+	CAN_DEVICE_DT_INST_DEFINE(inst,                                           \
+		&can_numaker_init,                                                \
+		NULL,                                                             \
+		&can_mcan_data_##inst,                                            \
+		&can_mcan_config_##inst,                                          \
+		POST_KERNEL,                                                      \
+		CONFIG_CAN_INIT_PRIORITY,                                         \
+		&can_numaker_driver_api);                                         \
+									          \
+	static void can_numaker_irq_config_func_##inst(const struct device *dev)  \
```
Comment (fabiobaltieri) [inline drivers/can/can_numaker.c:263]: You could move this before `can_numaker_config`, saves you from having to forward declare it.

File: drivers/can/can_numaker.c [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+		LOG_ERR("Invalid clock source rate index");
+		return -EIO;
+	}
+
+	LOG_DBG("Clock rate index/divider: %d/%d", clksrc_rate_idx, clkdiv_divider);
+
+	return 0;
+}
+
+static int can_numaker_init(const struct device *dev)
```
Comment (alexanderwachter) [inline drivers/can/can_numaker.c:74]: Just a hint:
You can put the section that needs the registers unlocked in a static inline function and call that from the init function. This way you can just return from the function in any error case without the goto.
It is also much safer, because there is no way to exit without locking.

```C
static inline int init_locked(...) {
    ret = do_something()
    if(ret) {
        return ret;
    }

    ret = do_something()
    if(ret) {
        return ret;
    }

}

int init(...)
{
    unlock_registers();
    ret = init_locked();
    lock_registers();
    return ret(
}
```

File: drivers/can/can_numaker.c (hunk 19 above)
Comment (cyliangtw) [reply drivers/can/can_numaker.c]: Yes, it could save the declaration.
Update at commit 764cfb094.

File: drivers/can/Kconfig.numaker (hunk 14 above)
[the remark this answers is not in the document]
Comment (henrikbrixandersen) [reply drivers/can/Kconfig.numaker:13]: I think you should remove it.

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
A few additional comments.

Please avoid resolving comments when addressing them, but rather leave it up to the reviewer to mark them as resolved if their concern was addressed.

File: drivers/can/can_numaker.c [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+	case 0:        /* HXT */
+		*rate = __HXT / clkdiv_divider;
+		break;
+	case 1:        /* PLLFOUT/2 */
+		*rate = (CLK_GetPLLClockFreq() / 2) / clkdiv_divider;
+		break;
+	case 2:        /* HCLK */
+		*rate = CLK_GetHCLKFreq() / clkdiv_divider;
+		break;
+	case 3:        /* HIRC */
```
[the remark this answers is not in the document]
Comment (cyliangtw) [reply drivers/can/can_numaker.c:61]: Yes, let it meaningful.

File: drivers/can/can_numaker.c [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+		LOG_ERR("Failed to initialize mcan: %d", rc);
+		return rc;
+	}
+
+	rc = can_numaker_get_core_clock(dev, &rate);
+	if (rc < 0) {
+		return rc;
+	}
+
+	LOG_DBG("CAN core clock: %d", rate);
```
[the remark this answers is not in the document]
Comment (cyliangtw) [reply drivers/can/can_numaker.c:126]: Done

File: drivers/can/can_numaker.c [hunk 23]
```diff
@@ -0,0 +1,286 @@
...
+{
+	const struct can_mcan_config *mcan_config = dev->config;
+	const struct can_numaker_config *config = mcan_config->custom;
+	int rc;
+
+	/* Validate this module's reset object */
+	if (!device_is_ready(config->reset.dev)) {
+		LOG_ERR("reset controller not ready");
+		return -ENODEV;
+	}
```
[the remark this answers is not in the document]
Comment (cyliangtw) [reply drivers/can/can_numaker.c]: Hi @henrikbrixandersen,
Agree with your above opinions and updated at commit  9d73702f0

**Review by henrikbrixandersen — APPROVED** [review]
Looks good! I assume all the CAN tests pass with hardware-in-the-loop testing on this board?

**Comment by cyliangtw** [comment]
> Looks good! I assume all the CAN tests pass with hardware-in-the-loop testing on this board?

Yes, I verified OK on this board before each git push.

**Review by fabiobaltieri — APPROVED** [review]
